src/Voron/Data/BTrees/Tree.MultiTree.cs

```diff
@@ -30,110 +31,304 @@ namespace Voron.Data.BTrees
             }
 
             var page = FindPageFor(key, out _);
-            if (page == null || page.LastMatch != 0)
+            if (page is not { LastMatch: 0 }) // Key is not in the tree. Use an optimized path for this scenario.
             {
-                MultiAddOnNewValue(key, value, maxNodeSize);
+                MultiBulkAddOnNewValue(key, values, maxNodeSize);
                 return;
             }
 
-            page = ModifyPage(page);
+            ContextBoundNativeList<int> newItems;
+            {
+                var readonlyKeyItem = page.GetNode(page.LastSearchPosition);
+                if (readonlyKeyItem->Flags == TreeNodeFlags.MultiValuePageRef)
+                {
 
-            var item = page.GetNode(page.LastSearchPosition);
+                    // We have an inner tree for the key. Since our values are sorted,
+                    // we can iterate over the elements and add them one by one.
+                    // As the page is cached, it is already loaded into memory. 
+                    // Therefore, this path could be further optimized by batching them into a single page.
+                    var existingTree = OpenMultiValueTree(key, readonlyKeyItem);
+                    for (int i = 0; i < values.Length; i++)
+                    {
+                        if (existingTree.Exists(values[i]))
+                            continue;
+                        
+                        existingTree.DirectAdd(values[i], 0,out _).Dispose();
+                    }
+                    return;
+                }
+                
+                if (readonlyKeyItem->Flags == TreeNodeFlags.PageRef)
+                    throw new InvalidOperationException("Multi trees don't use overflows");
+                
+                // We have a nested page here, which means our values are inlined into the page.
+                // Since we want to avoid modifying the page unless it is necessary,
+                // we will prepare a list of new terms based on the loaded data.
+                // This way, we can avoid modifying the page if we don't have to.
+                newItems = GetListOfNewItems(ref readonlyKeyItem, values);
+            }
 
-            // already was turned into a multi tree, not much to do here
-            if (item->Flags == TreeNodeFlags.MultiValuePageRef)
+            if (newItems.Count == 0)
             {
-                var existingTree = OpenMultiValueTree(key, item);
-                existingTree.DirectAdd(value, 0,out _).Dispose();
+                // Everything is already in the tree, nothing to do here.
+                newItems.Dispose();
                 return;
             }
 
-            if (item->Flags == TreeNodeFlags.PageRef)
-                throw new InvalidOperationException("Multi trees don't use overflows");
+            // We know there are new insertions. Let's try to insert them.
+            // Since we scoped the readonly keys in the previous scope,
```

Comment:
I don't follow this. What if the page is already modified, aren't you already reading from where you are writing?